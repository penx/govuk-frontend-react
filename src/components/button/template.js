// @flow

import * as React from 'react';
import type { ComponentType } from 'react';

import cx from 'classnames';
import styles from './_button.module.scss';

type Props = {
  className?: string,
  as?: ComponentType<{}> | string,
  type?: string,
  role?: string,
  disabled?: boolean,
  href?: string
};

/**
 *  Tracks https://github.com/alphagov/govuk-frontend/blob/master/src/components/button/template.njk
 */
const Button = ({ className, as, type, role, disabled, href, ...props }: Props) => {
  let Type = as || '';
  let computedType = type;
  let computedRole = role;
  let computedHref = href;
  if (Type === '') {
    if (href) {
      Type = 'a';
    } else {
      Type = 'button';
      if (!type) {
        computedType = 'submit';
      }
    }
  }

  if (Type === 'a') {
    computedHref = computedHref || '#';
    if (!role) {
      computedRole = 'button';
    }
  } else if (Type === 'input') {
    if (!type) {
      computedType = 'submit';
    }
  }
  return (
    <Type
      type={computedType}
      href={computedHref}
      role={computedRole}
      aria-disabled={disabled && 'true'}
      disabled={disabled && 'disabled'}
      className={cx(
        styles['govuk-button'],
        disabled && styles['govuk-button--disabled'],
        className
      )}
      {...props}
    />
  );
};

Button.defaultProps = {
  as: '',
  className: undefined,
  type: undefined,
  role: undefined,
  disabled: false,
  href: undefined
};

export default Button;
