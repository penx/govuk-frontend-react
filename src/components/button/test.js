import ReactDOM from 'react-dom';
import {
  render,
} from 'react-testing-library';
import 'jest-dom/extend-expect';

import defaultFixture, { hrefAndRole, buttonWithType } from './fixtures';

it('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(defaultFixture, div);
  ReactDOM.unmountComponentAtNode(div);
});

it('hrefAndRole', () => {
  const { getByText } = render(hrefAndRole);
  const button = getByText('HREF and role');
  expect(button).toHaveAttribute('href', '#test');
  expect(button).toHaveAttribute('role', 'presentation');
  expect(button.tagName).toBe('A');
});

it('buttonWithType', () => {
  const { getByText } = render(buttonWithType);
  const button = getByText('Button with type');
  expect(button).toHaveAttribute('type', 'button');
  expect(button.tagName).toBe('BUTTON');
});
