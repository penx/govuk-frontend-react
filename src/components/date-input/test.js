/* eslint-disable jsx-a11y/label-has-associated-control */

import React from 'react';
import { render } from 'react-testing-library';
import 'jest-dom/extend-expect';
import userEvent from 'user-event';

import DateInput from '.';

it('accepts hint prop as a string', () => {
  const { getByText } = render(<DateInput hint="Example Hint" />);

  expect(getByText('Example Hint')).toBeVisible();
});

it('accepts hint prop as a React element', () => {
  const { getByText } = render(<DateInput hint={<span>Example Hint</span>} />);

  expect(getByText('Example Hint')).toBeVisible();
});

it('accepts error message prop as a string', () => {
  const { getByText } = render(<DateInput errorMessage="Example Error" />);

  expect(getByText('Example Error')).toBeVisible();
});

it('accepts error message prop as a React element', () => {
  const { getByText } = render(<DateInput hint={<span>Example Error</span>} />);

  expect(getByText('Example Error')).toBeVisible();
});

it('accepts items with ids', () => {
  const { getByLabelText, getByTitle } = render(
    <form title="Form">
      <label htmlFor="my-element">Example</label>
      <DateInput
        items={[
          {
            name: 'day',
            placeholder: 'dd'
          },
          {
            name: 'test',
            id: 'my-element',
            placeholder: 'mm'
          },
          {
            name: 'year',
            placeholder: 'yyyy'
          }
        ]}
      />
    </form>
  );
  const form = getByTitle('Form');
  const monthField = getByLabelText('Example');
  userEvent.type(monthField, '1234');
  expect(form).toHaveFormValues({ test: 1234 });
});

it('accepts items without id and date-input has no id', () => {
  const { getByPlaceholderText, getByTitle } = render(
    <form title="Form">
      <DateInput
        items={[
          {
            name: 'day',
            placeholder: 'dd'
          },
          {
            name: 'test',
            placeholder: 'mm'
          },
          {
            name: 'year',
            placeholder: 'yyyy'
          }
        ]}
      />
    </form>
  );
  const form = getByTitle('Form');
  const monthField = getByPlaceholderText('mm');
  userEvent.type(monthField, '1234');
  expect(form).toHaveFormValues({ test: 1234 });
});

it('accepts items without id and no name and date-input has no id', () => {
  const { getByPlaceholderText, getByTitle } = render(
    <form title="Form">
      <DateInput
        items={[
          {
            placeholder: 'dd'
          },
          {
            placeholder: 'mm'
          },
          {
            placeholder: 'yyyy'
          }
        ]}
      />
    </form>
  );
  const form = getByTitle('Form');
  const monthField = getByPlaceholderText('mm');
  expect(monthField).not.toHaveAttribute('id');
});

it('shows a field error when supplied with a fieldErrors prop', () => {
  const { getByPlaceholderText } = render(
    <form title="Form">
      <DateInput
        fieldErrors={['month']}
        items={[
          {
            name: 'day',
            placeholder: 'dd'
          },
          {
            name: 'month',
            placeholder: 'mm'
          },
          {
            name: 'year',
            placeholder: 'yyyy'
          }
        ]}
      />
    </form>
  );
  const monthField = getByPlaceholderText('mm');
  expect(monthField).toHaveClass('govuk-input--error');
});
