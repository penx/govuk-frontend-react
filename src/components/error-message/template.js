// @flow

import * as React from 'react';
import cx from 'classnames';

export type ClassNames = {|
  'govuk-error-message': string
|};

const defaultClassNames: ClassNames = {
  'govuk-error-message': 'govuk-error-message'
};

const ErrorMessage = ({ className, classNames, ...props }: React.ElementProps<'span'>) => {
  const styles: ClassNames = {
    ...defaultClassNames,
    ...classNames
  };
  return <span {...props} className={cx(styles['govuk-error-message'], className)} />;
};

export default ErrorMessage;
