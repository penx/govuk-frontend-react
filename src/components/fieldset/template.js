// @flow

import * as React from 'react';
import cx from 'classnames';

type Props = React.ElementProps<'fieldset'> & {
  legend: {
    isPageHeading?: boolean,
    className: string,
    children: any
  }
};

export type ClassNames = {|
  'govuk-fieldset': string,
  'govuk-fieldset__legend': string,
  'govuk-fieldset__heading': string
|};

const defaultClassNames: ClassNames = {
  'govuk-fieldset': 'govuk-fieldset',
  'govuk-fieldset__legend': 'govuk-fieldset__legend',
  'govuk-fieldset__heading': 'govuk-fieldset__heading'
};

const Fieldset = ({ className, classNames, legend, children, describedBy, ...props }: Props) => {
  const styles: ClassNames = {
    ...defaultClassNames,
    ...classNames
  };
  return (
    <fieldset aria-describedby={describedBy} className={cx(styles['govuk-fieldset'], className)} {...props}>
      {legend && (
        <legend
          className={cx(
            styles['govuk-fieldset__legend'],
            legend.className &&
              legend.className
                .split(' ')
                // TODO: we can't unit test the `|| cn` branch due to the way we are mocking CSS modules
                .map(cn => styles[cn] || cn)
          )}
        >
          {legend.isPageHeading ? (
            <h1 className={styles['govuk-fieldset__heading']}>{legend.children}</h1>
          ) : (
            legend.children
          )}
        </legend>
      )}
      {children}
    </fieldset>
  );
};

export default Fieldset;
