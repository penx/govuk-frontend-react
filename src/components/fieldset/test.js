import React from 'react';
import { render } from 'react-testing-library';
import 'jest-dom/extend-expect';

import Fieldset from '.';

it('renders with a custom class on legend', () => {
  const { getByText } = render(<Fieldset legend={{ className: 'test', children: 'legend' }} />);
  expect(getByText('legend')).toHaveClass('test');
});

it('renders with a modifier class on legend', () => {
  const { getByText } = render(
    <Fieldset legend={{ className: 'govuk-fieldset__legend--xl', children: 'legend' }} />
  );
  expect(getByText('legend')).toHaveClass('govuk-fieldset__legend--xl');
});

it('renders with a legend as a page heading', () => {
  const { getByText } = render(<Fieldset legend={{ isPageHeading: true, children: 'legend' }} />);
  expect(getByText('legend')).toHaveClass('govuk-fieldset__heading');
});
