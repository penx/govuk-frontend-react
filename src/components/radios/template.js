// @flow

import * as React from 'react';
import cx from 'classnames';

import MergeFocus from 'merge-focus';

// import styles from './_radios.module.scss';

import ErrorMessage from '../error-message';
import Fieldset from '../fieldset';
import FormGroup from '../../objects/form-group';

import Item from './elements/item';

export type ClassNames = {|
  'govuk-radios__label': string,
  'govuk-radios__hint': string,
  'govuk-radios__conditional': string,
  'govuk-radios__conditional--hidden': string,
  'govuk-radios__divider': string,
  'govuk-radios__item': string,
  'govuk-radios__input': string,
  'govuk-radios': string,
  'govuk-radios--conditional': string
|};

const defaultClassNames: ClassNames = {
  'govuk-radios__label': 'govuk-radios__label',
  'govuk-radios__hint': 'govuk-radios__hint',
  'govuk-radios__conditional': 'govuk-radios__conditional',
  'govuk-radios__conditional--hidden': 'govuk-radios__conditional--hidden',
  'govuk-radios__divider': 'govuk-radios__divider',
  'govuk-radios__item': 'govuk-radios__item',
  'govuk-radios__input': 'govuk-radios__input',
  'govuk-radios': 'govuk-radios',
  'govuk-radios--conditional': 'govuk-radios--conditional'
};

const styles = defaultClassNames;

const Radios = ({
  name,
  idPrefix = name,
  className,
  items = [],
  hint,
  errorMessage,
  children,
  id,
  formGroup,
  fieldset,
  onBlur,
  onFocus,
  value,
  classNames,
  elements,
  components,
  ...props
}: React.ElementProps<'label'>) => {
  let describedBy;
  if (hint) {
    describedBy = `${idPrefix}-hint`;
  }
  if (errorMessage) {
    describedBy = describedBy ? `${describedBy} ${idPrefix}-error` : `${idPrefix}-error`;
  }
  const isConditional = items.some(item => item.conditional);

  const inner = (
    <>
      {hint && <Hint id={`${idPrefix}-hint`} {...hint} />}
      {errorMessage && (
        <ErrorMessage
          id={`${idPrefix}-error`}
          {...(typeof errorMessage !== 'object' || React.isValidElement(errorMessage)
            ? { children: errorMessage }
            : errorMessage)}
        />
      )}
      <div
        className={cx(
          styles['govuk-radios'],
          className,
          isConditional && styles['govuk-radios--conditional']
        )}
        {...props}
        data-module={isConditional && 'radios'}
      >
        {/*
          TODO: merge values alternative for radios?
          e.g.
          https://github.com/mui-org/material-ui/blob/master/packages/material-ui/src/RadioGroup/RadioGroup.js
          https://github.com/IBM/carbon-components-react/blob/master/src/components/RadioButtonGroup/RadioButtonGroup.js
          https://github.com/auth0/cosmos/blob/master/core/components/atoms/radio/radio.js
          */}
        {children || (
          <MergeFocus
            onBlur={onBlur}
            onFocus={onFocus}
            keys={items.map((item, i) => item.id || (id ? `${id}-${i + 1}` : i + 1))}
          >
            {mergeFocus => (
              <>
                {items.map((item, i) => {
                  const index = i + 1; // using 1-index to match govuk-frontend
                  const key = item.id || (id ? `${id}-${index}` : index);
                  const itemId = item.id || (idPrefix ? `${idPrefix}-${index}` : `${id}-${index}`);
                  return (
                    <Radios.Item name={name} key={key} {...item} id={itemId} {...mergeFocus[key]} />
                  );
                })}
              </>
            )}
          </MergeFocus>
        )}
      </div>
    </>
  );

  return (
    <FormGroup error={!!errorMessage} className={formGroup && formGroup.className}>
      {fieldset ? (
        <Fieldset aria-describedby={describedBy} {...fieldset}>
          {inner}
        </Fieldset>
      ) : (
        inner
      )}
    </FormGroup>
  );
};

Radios.Item = Item;
Radios.Conditional = RadioConditional;
Radios.Hint = RadioHint;
Radios.Label = RadioLabel;

Radios.defaultProps = {
  name: undefined,
  idPrefix: undefined,
  className: undefined,
  items: [],
  hint: undefined,
  errorMessage: undefined,
  children: undefined,
  id: undefined,
  formGroup: undefined,
  fieldset: undefined,
  onBlur: undefined,
  onFocus: undefined,
  value: undefined,
  classNames: {},
  elements: {},
  components: {}
};

export default Radios;
