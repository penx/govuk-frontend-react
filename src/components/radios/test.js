// id but no item.id
import * as React from 'react';
import { render } from 'react-testing-library';
import userEvent from 'user-event';
import Radios from '.';

it('accepts items without id and radios has id', () => {
  const { getByLabelText, getByTitle } = render(
    <form title="Form">
      <Radios
        id="foo"
        name="bar"
        items={[
          {
            value: 'y',
            children: 'yarp'
          },
          {
            value: 'n',
            children: 'nope'
          },
          {
            value: 'm',
            children: 'mabes'
          }
        ]}
      />
    </form>
  );
  const form = getByTitle('Form');
  const noRadio = getByLabelText('nope');
  userEvent.click(noRadio);
  expect(form).toHaveFormValues({ bar: 'n' });
});
