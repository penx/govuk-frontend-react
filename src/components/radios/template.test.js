require('govuk-frontend-template-spec/components/radios/template.test');
