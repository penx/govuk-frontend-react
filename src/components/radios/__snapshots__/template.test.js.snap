// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Radios nested dependant components passes through fieldset params without breaking 1`] = `

<fieldset class="govuk-fieldset app-fieldset--custom-modifier"
          aria-describedby="example-hint example-error"
          data-attribute="value"
          data-second-attribute="second-value"
>
  <legend class="govuk-fieldset__legend">
    Have you changed your name?
  </legend>
</fieldset>

`;

exports[`Radios nested dependant components passes through html fieldset params without breaking 1`] = `

<fieldset class="govuk-fieldset">
  <legend class="govuk-fieldset__legend">
    Have
    <b>
      you
    </b>
    changed your name?
  </legend>
</fieldset>

`;

exports[`Radios nested dependant components passes through label params without breaking 1`] = `

<label class="govuk-label govuk-radios__label"
       data-attribute="value"
       data-second-attribute="second-value"
       for="example-name-1"
>
  <b>
    Yes
  </b>
</label>
<label class="govuk-label govuk-radios__label"
       for="example-name-2"
>
  &lt;b&gt;No&lt;/b&gt;
</label>

`;

exports[`Radios when they include a hint renders the hint 1`] = `

<span id="example-hint"
      class="govuk-hint"
>
  This includes changing your last name or spelling your name differently.
</span>

`;

exports[`Radios when they include an error message renders the error message 1`] = `

<span id="example-error"
      class="govuk-error-message"
>
  Please select an option
</span>

`;
